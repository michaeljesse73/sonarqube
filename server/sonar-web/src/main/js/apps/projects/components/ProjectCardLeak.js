/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import classNames from 'classnames';
import moment from 'moment';
import { Link } from 'react-router';
import ProjectCardQualityGate from './ProjectCardQualityGate';
import ProjectCardLeakMeasures from './ProjectCardLeakMeasures';
import FavoriteContainer from '../../../components/controls/FavoriteContainer';
import Organization from '../../../components/shared/Organization';
import TagsList from '../../../components/tags/TagsList';
import PrivateBadge from '../../../components/common/PrivateBadge';
import { translate, translateWithParameters } from '../../../helpers/l10n';

type Props = {
  measures: { [string]: string },
  organization?: { key: string },
  project?: {
    analysisDate?: string,
    key: string,
    leakPeriodDate?: string,
    name: string,
    tags: Array<string>,
    isFavorite?: boolean,
    organization?: string,
    visibility?: boolean
  }
};

export default function ProjectCardLeak({ measures, organization, project }: Props) {
  if (project == null) {
    return null;
  }

  const isProjectAnalyzed = project.analysisDate != null;
  const isPrivate = project.visibility === 'private';
  const hasLeakPeriodStart = project.leakPeriodDate != null;
  const hasTags = project.tags.length > 0;
  const showOrganization = organization == null && project.organization != null;

  // check for particular measures because only some measures can be loaded
  // if coming from visualizations tab
  const areProjectMeasuresLoaded = measures != null && measures['new_bugs'];

  const displayQualityGate = areProjectMeasuresLoaded && isProjectAnalyzed;
  const className = classNames('boxed-group', 'project-card', {
    'boxed-group-loading': isProjectAnalyzed && hasLeakPeriodStart && !areProjectMeasuresLoaded
  });

  return (
    <div data-key={project.key} className={className}>
      <div className="boxed-group-header clearfix">
        {project.isFavorite != null &&
          <FavoriteContainer className="spacer-right" componentKey={project.key} />}
        <h2 className="project-card-name">
          {showOrganization &&
            <span className="text-normal">
              <Organization organizationKey={project.organization} />
            </span>}
          <Link to={{ pathname: '/dashboard', query: { id: project.key } }}>{project.name}</Link>
        </h2>
        {displayQualityGate && <ProjectCardQualityGate status={measures['alert_status']} />}
        <div className="pull-right text-right">
          {isPrivate && <PrivateBadge className="spacer-left" tooltipPlacement="left" />}
          {hasTags && <TagsList tags={project.tags} customClass="spacer-left" />}
        </div>
        {isProjectAnalyzed &&
          hasLeakPeriodStart &&
          <div className="project-card-dates note text-right pull-right">
            {hasLeakPeriodStart &&
              <span className="project-card-leak-date pull-right">
                {translateWithParameters(
                  'projects.leak_period_x',
                  moment(project.leakPeriodDate).fromNow()
                )}
              </span>}
            {isProjectAnalyzed &&
              <span>
                {translateWithParameters(
                  'projects.last_analysis_on_x',
                  moment(project.analysisDate).format('LLL')
                )}
              </span>}
          </div>}
      </div>

      {isProjectAnalyzed && hasLeakPeriodStart
        ? <div className="boxed-group-inner">
            {areProjectMeasuresLoaded && <ProjectCardLeakMeasures measures={measures} />}
          </div>
        : <div className="boxed-group-inner">
            <div className="note project-card-not-analyzed">
              {isProjectAnalyzed
                ? translate('projects.no_leak_period')
                : translate('projects.not_analyzed')}
            </div>
          </div>}
    </div>
  );
}
