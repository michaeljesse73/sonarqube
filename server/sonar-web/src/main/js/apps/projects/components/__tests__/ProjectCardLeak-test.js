/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import ProjectCardLeak from '../ProjectCardLeak';

const PROJECT = {
  analysisDate: '2017-01-01',
  leakPeriodDate: '2016-12-01',
  key: 'foo',
  name: 'Foo',
  tags: []
};
const MEASURES = {
  alert_status: 'OK',
  reliability_rating: '1.0',
  sqale_rating: '1.0',
  new_bugs: 12
};

jest.mock('moment', () => () => ({
  format: () => 'March 1, 2017 9:36 AM',
  fromNow: () => 'a month ago'
}));

it('should display analysis date and leak start date', () => {
  const card = shallow(<ProjectCardLeak type="leak" measures={MEASURES} project={PROJECT} />);
  expect(card.find('.project-card-dates').exists()).toBeTruthy();
  expect(card.find('.project-card-dates').find('span').getNodes()).toHaveLength(2);
});

it('should not display analysis date or leak start date', () => {
  const project = { ...PROJECT, analysisDate: undefined };
  const card = shallow(<ProjectCardLeak type="leak" measures={MEASURES} project={project} />);
  expect(card.find('.project-card-dates').exists()).toBeFalsy();
});

it('should display loading', () => {
  const measures = { ...MEASURES, new_bugs: undefined };
  expect(
    shallow(<ProjectCardLeak type="leak" measures={measures} project={PROJECT} />)
      .find('.boxed-group')
      .hasClass('boxed-group-loading')
  ).toBeTruthy();
});

it('should display tags', () => {
  const project = { ...PROJECT, tags: ['foo', 'bar'] };
  expect(
    shallow(<ProjectCardLeak type="leak" project={project} />).find('TagsList').exists()
  ).toBeTruthy();
});

it('should private badge', () => {
  const project = { ...PROJECT, visibility: 'private' };
  expect(
    shallow(<ProjectCardLeak type="leak" project={project} />).find('PrivateBadge').exists()
  ).toBeTruthy();
});

it('should display the leak measures and quality gate', () => {
  expect(
    shallow(<ProjectCardLeak type="leak" measures={MEASURES} project={PROJECT} />)
  ).toMatchSnapshot();
});
