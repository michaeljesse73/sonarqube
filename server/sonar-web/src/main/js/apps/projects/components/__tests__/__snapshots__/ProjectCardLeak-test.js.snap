// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the leak measures and quality gate 1`] = `
<div
  className="boxed-group project-card"
  data-key="foo"
>
  <div
    className="boxed-group-header clearfix"
  >
    <h2
      className="project-card-name"
    >
      <Link
        onlyActiveOnIndex={false}
        style={Object {}}
        to={
          Object {
            "pathname": "/dashboard",
            "query": Object {
              "id": "foo",
            },
          }
        }
      >
        Foo
      </Link>
    </h2>
    <ProjectCardQualityGate
      status="OK"
    />
    <div
      className="pull-right text-right"
    />
    <div
      className="project-card-dates note text-right pull-right"
    >
      <span
        className="project-card-leak-date pull-right"
      >
        projects.leak_period_x.a month ago
      </span>
      <span>
        projects.last_analysis_on_x.March 1, 2017 9:36 AM
      </span>
    </div>
  </div>
  <div
    className="boxed-group-inner"
  >
    <ProjectCardLeakMeasures
      measures={
        Object {
          "alert_status": "OK",
          "new_bugs": 12,
          "reliability_rating": "1.0",
          "sqale_rating": "1.0",
        }
      }
    />
  </div>
</div>
`;
